### PR TITLE
Rollback callback

### PR DESCRIPTION
**Pull request by yash620**
This adds in a framework for having callbacks when a transaction rollsback.

### CODE REVIEW

File: src/include/transaction/transaction_context.h [hunk 1]
```diff
@@ -129,5 +147,8 @@ class TransactionContext {
   // log manager will set this to be true when log records are processed (not necessarily flushed, but will not be read
   // again in the future), so it can be garbage-collected safely.
   bool log_processed_ = false;
+
+  // Functions to be invoked on rollback
+  std::map<transaction::rollback_fn, void *> rollback_functions_;
```
Comment (mbutrovich) [inline src/include/transaction/transaction_context.h]: When @jrolli mentioned this to me the other day it sounded like a stack was what we were going to use. Do we need ordered and random element access that a map provides?

Comment (apavlo) [reply src/include/transaction/transaction_context.h]: We don't need random access. AFAIK all constraint checks are commutative, so the order does not matter.

Comment (mbutrovich) [reply src/include/transaction/transaction_context.h]: Suggest replacing with std::forward_list then.

**Comment by mbutrovich** [comment]
Once this is in, we should discuss if this is the way we want to approach handling index rollbacks as well, and if so modify the index wrapper as needed.

File: src/include/transaction/transaction_context.h (hunk 1 above)
Comment (yash620) [reply src/include/transaction/transaction_context.h]: Alright, I'll go ahead and do that

**Comment by yash620** [comment]
John's framework of callbacks will take care of this. No need for this anymore.

**Pull request closed without merging**